### ISSUE TITLE
Refined `textLine` Matching

### ISSUE THREAD

**Issue by joel-duffie** [nut-tree/nut.js#552]
**Short overview**
Improve the textLine search method. Add the option to only return the `Region` around the string of words given, when using `partialMatch`.

**Use case**
When searching for a `textLine` with `partialMatch` set to `true`, there should be an option to limit the returned `Region` to only the portion of the text line which includes the given string.

**Detailed Description** 
I have a UI element I desire to locate. It's text reads "Drag Action To Cue In Presentation".

![image](https://github.com/nut-tree/nut.js/assets/141655392/878b4811-495d-4655-8254-21e8c5767463)

I'm using `textLine`, as it is the only OCR option to search for multi-word strings. Here's a snippet of the code I'm using.

```js
        ocrParams = {
            lang: [ocr.Language.English],
            partialMatch: true,
            caseSensitive: false
        }
        nut.screen.config.autoHighlight = true;
        nut.screen.config.highlightDurationMs = 3000;
        nut.screen.config.confidence = 0.85;
        const apTitle = await nut.screen.find(nut.textLine('Drag Action To Cue In Presentation'),
            {providerData: ocrParams});
```

I think it's obvious that I'm only searching for the words "Drag Action To Cue In Presentation".
However, this is the region found currently using  the OCR plugin. It's seeing all the other text in the text line as 'part' of the region, no matter how far away the text is - it even sees the `x` on the close button as a part of that line.

![ocr_textLine_error](https://github.com/nut-tree/nut.js/assets/141655392/40d0c46e-0c03-4a10-926d-748e34c82721)

Here's the error I get if I raise `confidence` slightly higher.

```
Error: [nut.js] - Error: Searching for line-query-Drag Action To Cue In Presentation failed.
Reason: 'Error: OCR run failed. Reason: Error: Failed to detect text line drag action to cue in presentation
    with config {"lang":["eng"],"partialMatch":true,"caseSensitive":false}. 
Best match is Drag Action To Cue In Presentation x DefaulTimer with confidence 0.8522521209716797'
```

As you can see, the matched line is _"Drag Action To Cue In Presentation x DefaulTimer"_. But I only desire the `Region` containing _"Drag Action To Cue In Presentation"_.

**To be clear**, I understand why all the words on the screen on the same y-level are matched as part of the "text line".
**What I propose** is that the `Region` returned by `textLine` be limited to only the given string's words, in this case, "Drag Action To Cue In Presentation". This could be included as a parameter for the OCR search. At this time, we cannot confidently use OCR for multi-word lines because of this issue, which could prove to be a big deal for us down the road.

**Comment by s1hofmann** [comment]
Hi @joel-duffie 👋 

Thanks for your suggestion, it's a nice improvement and considered as one of the next work items.

**Comment by s1hofmann** [comment]
@joel-duffie Can you give `@nut-tree/plugin-ocr@next` a try?

**Comment by joel-duffie** [comment]
@s1hofmann

I gave it a try, and it seems to work!

There was an error I ran into, however, that I figured you'd want to know about.

I ran this code...
```js
nut.screen.config.confidence = 0.80;
nut.screen.config.autoHighlight = true;
nut.screen.config.highlightDurationMs = 3000;

const line = await nut.screen.find(nut.textLine('Drag Action To Cue In Presentation'), {
    providerData: {
        lang: ocr.Language.English,
        partialMatch: true,
        caseSensitive: true
    }
});
```

...and got this error:

```
Error: [nut.js] - Error: Searching for line-query-Drag Action To Cue In Presentation failed. Reason: 'Error: OCR run failed. Reason: TypeError: languages.join is not a function'
    at ScreenClass.find (C:\Users\-----\nut-js-testing\node_modules\@nut-tree\nut-js\dist\lib\screen.class.js:169:27)
    at async C:\Users\-----\nut-js-testing\index.js:22:22
```

I didn't receive the error, and it works, when changed to this:

```js
nut.screen.config.confidence = 0.80;
nut.screen.config.autoHighlight = true;
nut.screen.config.highlightDurationMs = 3000;

const line = await nut.screen.find(nut.textLine('Drag Action To Cue In Presentation'));
```

**Comment by s1hofmann** [comment]
Yes, because the `lang` parameter is supposed to be an array, which the default is.
But I can add a check for it.

**Comment by joel-duffie** [comment]
Oops - sorry about that

**Comment by s1hofmann** [comment]
No worries 😃

If the issue itself is solved I’ll keep it open until the next release 👍

**Comment by joel-duffie** [comment]
Hi @s1hofmann,
Is it possible that the changes made in the `next` ocr version may have caused ocr in context menus to stop working?
I'm experiencing an inability to match words in the menubar menus (under File, View, etc). I am still able to image-match on these menus.

**Comment by s1hofmann** [comment]
@joel-duffie The recognition code didn’t change, it only affected the postprocessing of results. 
Is it reproducible when switching between versions?

**Comment by joel-duffie** [comment]
@s1hofmann 
I downgraded to OCR version 1.0.2, and there is no change in behavior.

Here's a simple example:

In my application, I open the 'View' menu item in the menu bar (found by OCR).
In the menu underneath, I search for the term 'Calendar'. Here's the code:
```js
let reg = await nut.screen.find(nut.singleWord('Calendar'), {
    confidence: 0.70,
    providerData: {
        lang: [ocr.Language.English],
        partialMatch: true,
        matchCase: false
    }
});
```

However, there are no results found:

```
Error: [nut.js] - Error: Searching for word-query-Calendar failed. 
Reason: 'Error: OCR run failed. 
Reason: Error: Failed to detect word calendar with config {"lang":["eng"],"partialMatch":true,"matchCase":false}'
    at ScreenClass.find (C:\Users\---\nut-js-testing\node_modules\@nut-tree\nut-js\dist\lib\screen.class.js:169:27)
    at async C:\Users\---\nut-js-testing\ocr-testing.js:31:15
```

I can open a bug on github if you'd like.

**Comment by s1hofmann** [comment]
@joel-duffie The menu on your screenshot has not that much contrast, so this might become a problem.

Two suggestions you could try:

1. Disable preprocessing and use the raw input by passing this `preprocessConfig`: 
```js
let reg = await nut.screen.find(nut.singleWord('Calendar'), {
    confidence: 0.70,
    providerData: {
        lang: [ocr.Language.English],
        partialMatch: true,
        matchCase: false,
        preprocessConfig: {
            binarize: false
        }
    }
});
```
2. Make use of keyboard shortcuts. Depends on your task at hand, but the `Calendar` entry has a shortcut defined you could use to execute the action

**Comment by s1hofmann** [comment]
@joel-duffie One more thing to do is to limit the search region to the current active window, which would be the open menu in your case.

```js
const activeWindowRegion = () => getActiveWindow().then((window) => window.region);
const location = await screen.find(singleWord("Calendar"), { searchRegion: activeWindowRegion()});
```

**Comment by s1hofmann** [comment]
Closed in OCR plugin release v2.0.0
